File: frontend/herdr-web/src/demo/demoData.ts
/**
 * Demo fixtures — port of herdr-harness-ios Models/DemoData.swift (the 3
 * workspaces, tabs, panes, 2 alerts). Shapes follow src/types/herdr.ts
 * (the web API contract) rather than the Swift models: no `tokens`, no
 * `title` (pane display comes from terminal_title), agents arrays are
 * empty (nothing on the web renders them).
 *
 * w1:p1 (Codex) carries a Pi semantic capability so the pane offers Chat;
 * the demo Pi snapshot (below) reports `available: false`, which lands the
 * chat in its "native transcript unavailable" empty state instead of
 * streaming.
 */

import type {
  Alert,
  AlertsResponse,
  Health,
  Pane,
  Workspace,
  WorkspacesResponse,
} from "../types/herdr";

export const DEMO_GENERATED_AT = "2030-01-01T12:00:00Z";

const DEMO_CWD = "/tmp/herdr-demo";

function pane(
  paneId: string,
  tabId: string,
  status: Pane["agent_status"],
  title: string,
  agent: string | null,
  revision: number,
  piSemantic?: Pane["pi_semantic"],
): Pane {
  return {
    pane_id: paneId,
    terminal_id: `term_${paneId.replace(":", "_")}`,
    workspace_id: paneId.split(":")[0] ?? "",
    tab_id: tabId,
    focused: paneId === "w1:p1",
    cwd: DEMO_CWD,
    foreground_cwd: "",
    terminal_title: title,
    terminal_title_stripped: title,
    ...(agent !== null ? { agent } : {}),
    agent_status: status,
    scroll: { offset_from_bottom: 0, max_offset_from_bottom: 0, viewport_rows: 26 },
    revision,
    ...(piSemantic !== undefined ? { pi_semantic: piSemantic } : {}),
  };
}

/**
 * Swift demo tab rule: sorted tab ids, first tab "Agents", the rest
 * "Tests"; per-tab status mirrors the workspace status.
 */
function tabs(workspaceId: string, tabIds: string[], status: Workspace["agent_status"]) {
  return tabIds.map((tabId, index) => {
    const paneCount = workspacePanes.filter((candidate) => candidate.tab_id === tabId).length;
    return {
      tab_id: tabId,
      workspace_id: workspaceId,
      number: index + 1,
      label: index === 0 ? "Agents" : "Tests",
      focused: index === 0,
      pane_count: paneCount,
      agent_status: status,
    };
  });
}

// `tabs()` above counts panes across all demo workspaces by tab id; tab ids
// are workspace-scoped (`w1:t1`), so a global filter is correct.
const workspacePanes: Pane[] = [
  pane("w1:p1", "w1:t1", "working", "Plan a fictional herb garden", "codex", 184, {
    available: true,
    connected: false,
    protocol_version: 1,
    session_id: "demo-pi-session",
    cursor: 0,
    oldest_cursor: 0,
    capabilities: {
      prompt: true,
      steer: true,
      followUp: true,
      abort: true,
      listModels: false,
      setModel: false,
      setThinkingLevel: false,
      interactionResponse: false,
    },
    generated_at: DEMO_GENERATED_AT,
  }),
  pane("w1:p2", "w1:t1", "blocked", "Choose sample garden colors", "claude", 97),
  pane("w1:p3", "w1:t2", "unknown", "Unit tests", null, 52),
  pane("w2:p1", "w2:t1", "done", "Sample reading list export", "codex", 311),
  pane("w2:p2", "w2:t1", "working", "Validate the example book list", "claude", 118),
  pane("w3:p1", "w3:t1", "idle", "Describe the sample weather display", "codex", 44),
];

/** Swift demo layout rule: 120×36 area, panes tiled left to right. */
function layout(workspaceId: string, tabId: string, paneIds: string[]) {
  const width = 120;
  const paneWidth = width / Math.max(paneIds.length, 1);
  return {
    workspace_id: workspaceId,
    tab_id: tabId,
    zoomed: false,
    area: { x: 0, y: 0, width, height: 36 },
    focused_pane_id: paneIds[0] ?? "",
    panes: paneIds.map((paneId, index) => ({
      pane_id: paneId,
      focused: index === 0,
      rect: { x: index * paneWidth, y: 0, width: paneWidth, height: 36 },
    })),
    splits: [],
  };
}

function workspace(
  workspaceId: string,
  number: number,
  label: string,
  path: string,
  status: Workspace["agent_status"],
  tabIds: string[],
  layoutPaneIds: string[][],
): Workspace {
  const panes = workspacePanes.filter((candidate) => candidate.workspace_id === workspaceId);
  const activeTabId = panes[0]?.tab_id ?? "";
  return {
    workspace_id: workspaceId,
    number,
    label,
    focused: number === 1,
    pane_count: panes.length,
    tab_count: tabIds.length,
    active_tab_id: activeTabId,
    agent_status: status,
    tabs: tabs(workspaceId, tabIds, status),
    panes,
    agents: [],
    layouts: layoutPaneIds.map((paneIds) => layout(workspaceId, activeTabId, paneIds)),
    worktree: {
      repo_key: label.toLowerCase().replace(/ /g, "-"),
      repo_name: label,
      repo_root: path,
      checkout_path: path,
      is_linked_worktree: number === 1,
    },
  };
}

/** Swift `DemoData.workspaces` — attention-ranked workspace status per pane. */
export const demoWorkspaces: Workspace[] = [
  workspace("w1", 1, "Garden Planner", "/tmp/herdr-demo/garden-planner", "blocked", ["w1:t1", "w1:t2"], [
    ["w1:p1", "w1:p2"],
  ]),
  workspace("w2", 2, "Reading Journal", "/tmp/herdr-demo/reading-journal", "done", ["w2:t1"], [
    ["w2:p1", "w2:p2"],
  ]),
  workspace("w3", 3, "Weather Station", "/tmp/herdr-demo/weather-station", "idle", ["w3:t1"], [
    ["w3:p1"],
  ]),
];

/** Swift `DemoData.alerts`, mapped onto the web Alert shape. */
export const demoAlerts: Alert[] = [
  {
    id: "demo-blocked",
    kind: "agent_blocked",
    status: "blocked",
    previousStatus: "working",
    severity: "warning",
    title: "Choose the example garden theme",
    message: "The demo planning agent is waiting for a garden color choice.",
    workspaceId: "w1",
    workspaceLabel: "Garden Planner",
    tabId: "w1:t1",
    tabLabel: "Agents",
    paneId: "w1:p2",
    paneTitle: "Choose sample garden colors",
    agentName: "claude",
    createdAt: "2030-01-01T11:58:00Z",
    isRead: false,
    readAt: null,
    action: { type: "open_pane", paneId: "w1:p2" },
  },
  {
    id: "demo-done",
    kind: "agent_done",
    status: "done",
    previousStatus: "working",
    severity: "success",
    title: "The sample book list is ready",
    message: "The demo reading list export contains three fictional book entries.",
    workspaceId: "w2",
    workspaceLabel: "Reading Journal",
    tabId: "w2:t1",
    tabLabel: "Agents",
    paneId: "w2:p1",
    paneTitle: "Sample reading list export",
    agentName: "codex",
    createdAt: "2030-01-01T11:55:00Z",
    isRead: false,
    readAt: null,
    action: { type: "open_pane", paneId: "w2:p1" },
  },
];

/** Demo /health: a healthy, fully attached herdr. */
export const demoHealth: Health = {
  ok: true,
  service: "herdr-harness",
  session: "default",
  herdr: {
    connected: true,
    requestConnected: true,
    eventsConnected: true,
    socketFound: true,
    version: "0.8.0",
    protocol: 19,
    lastError: null,
  },
  cache: { available: true, stale: false, generatedAt: DEMO_GENERATED_AT },
  alerts: { unread: demoAlerts.length },
  generatedAt: DEMO_GENERATED_AT,
};

export function demoWorkspacesResponse(alerts: Alert[]): WorkspacesResponse {
  return { ok: true, workspaces: demoWorkspaces, alerts, generatedAt: DEMO_GENERATED_AT };
}

export function demoAlertsResponse(alerts: Alert[]): AlertsResponse {
  return {
    ok: true,
    alerts,
    unreadCount: alerts.filter((alert) => !alert.isRead).length,
    generatedAt: DEMO_GENERATED_AT,
  };
}

export function demoWorkspaceResponse(workspaceId: string): Workspace | undefined {
  return demoWorkspaces.find((candidate) => candidate.workspace_id === workspaceId);
}

/**
 * Demo terminal snapshot: always empty. The plan keeps demo terminals
 * streamless — "No terminal output yet." is the expected rendering.
 */
export function demoTerminalOutputResponse(paneId: string): unknown {
  return {
    ok: true,
    output: {
      pane_id: paneId,
      workspace_id: paneId.split(":")[0] ?? "",
      tab_id: "",
      source: "recent_unwrapped",
      format: "text",
      text: "",
      revision: 0,
      truncated: false,
    },
    result: null,
    generatedAt: DEMO_GENERATED_AT,
  };
}

/**
 * Demo Pi snapshot: `available: false` → the follow loop parks in the
 * "native transcript unavailable" state (empty chat, no streaming).
 */
export function demoPiSnapshotResponse(paneId: string): unknown {
  return {
    ok: true,
    protocol: { name: "herdr.pi.semantic", version: 1 },
    paneId,
    available: false,
    connected: false,
    entries: [],
    cursor: null,
    oldestCursor: null,
    truncated: false,
    generatedAt: DEMO_GENERATED_AT,
  };
}

File: herdr-harness-ios/herdr-harness-ios/Models/DemoData.swift
import Foundation

/// Entirely fictional garden, reading, and weather examples; never captured user work.
enum DemoData {
    static let workspaces: [HerdrWorkspace] = [
        workspace(
            id: "w1",
            number: 1,
            label: "Garden Planner",
            path: "/tmp/herdr-demo/garden-planner",
            panes: [
                pane(id: "w1:p1", tabID: "w1:t1", status: .working, title: "Plan a fictional herb garden", agent: "Codex", revision: 184, firstSeenAt: Date().addingTimeInterval(-3_600), lastActivityAt: Date(), workingSince: Date().addingTimeInterval(-47 * 60)),
                pane(id: "w1:p2", tabID: "w1:t1", status: .blocked, title: "Choose sample garden colors", agent: "Claude", revision: 97, firstSeenAt: Date().addingTimeInterval(-7_200), lastActivityAt: Date().addingTimeInterval(-22 * 60)),
                pane(id: "w1:p3", tabID: "w1:t2", status: .unknown, title: "Unit tests", agent: nil, revision: 52, firstSeenAt: Date().addingTimeInterval(-3 * 86_400), lastActivityAt: Date().addingTimeInterval(-2 * 86_400)),
            ],
            layouts: [layout(workspaceID: "w1", tabID: "w1:t1", paneIDs: ["w1:p1", "w1:p2"])]
        ),
        workspace(
            id: "w2",
            number: 2,
            label: "Reading Journal",
            path: "/tmp/herdr-demo/reading-journal",
            panes: [
                pane(id: "w2:p1", tabID: "w2:t1", status: .done, title: "Sample reading list export", agent: "Codex", revision: 311, firstSeenAt: Date(), lastActivityAt: Date()),
                pane(id: "w2:p2", tabID: "w2:t1", status: .working, title: "Validate the example book list", agent: "Claude", revision: 118, firstSeenAt: Date().addingTimeInterval(-5_400), lastActivityAt: Date(), workingSince: Date().addingTimeInterval(-35 * 60)),
            ],
            layouts: [layout(workspaceID: "w2", tabID: "w2:t1", paneIDs: ["w2:p1", "w2:p2"])]
        ),
        workspace(
            id: "w3",
            number: 3,
            label: "Weather Station",
            path: "/tmp/herdr-demo/weather-station",
            panes: [
                pane(id: "w3:p1", tabID: "w3:t1", status: .idle, title: "Describe the sample weather display", agent: "Codex", revision: 44, firstSeenAt: Date().addingTimeInterval(-10 * 86_400), lastActivityAt: Date().addingTimeInterval(-8 * 86_400)),
            ],
            layouts: [layout(workspaceID: "w3", tabID: "w3:t1", paneIDs: ["w3:p1"])]
        ),
    ]

    static let alerts: [HerdrAlert] = [
        HerdrAlert(
            id: "demo-blocked",
            workspaceID: "w1",
            paneID: "w1:p2",
            status: .blocked,
            title: "Choose the example garden theme",
            message: "The demo planning agent is waiting for a garden color choice.",
            createdAt: "2030-01-01T11:58:00Z",
            isRead: false
        ),
        HerdrAlert(
            id: "demo-done",
            workspaceID: "w2",
            paneID: "w2:p1",
            status: .done,
            title: "The sample book list is ready",
            message: "The demo reading list export contains three fictional book entries.",
            createdAt: "2030-01-01T11:55:00Z",
            isRead: false
        ),
    ]

    static let secondaryWorkspaces: [HerdrWorkspace] = [
        workspace(
            id: "w1",
            number: 1,
            label: "Art Notebook",
            path: "/tmp/herdr-demo/art-notebook",
            panes: [
                pane(id: "w1:p1", tabID: "w1:t1", status: .working, title: "Sketch a fictional postcard", agent: "Codex", revision: 28),
                pane(id: "w1:p2", tabID: "w1:t1", status: .idle, title: "Choose the sample postcard border", agent: "Claude", revision: 12),
            ],
            layouts: [layout(workspaceID: "w1", tabID: "w1:t1", paneIDs: ["w1:p1", "w1:p2"])]
        ),
        workspace(
            id: "w2",
            number: 2,
            label: "Weather Samples",
            path: "/tmp/herdr-demo/weather-samples",
            panes: [
                pane(id: "w2:p1", tabID: "w2:t1", status: .done, title: "Validate fictional weather readings", agent: "Codex", revision: 7),
            ],
            layouts: [layout(workspaceID: "w2", tabID: "w2:t1", paneIDs: ["w2:p1"])]
        ),
    ]

    static func terminalText(for paneID: String) -> String {
        switch paneID {
        case "demo1|w1:p2":
            """
            Fictional Garden Planner demo

            Choose a color for the example herb-garden tiles:
              1. Sage green
              2. Marigold yellow

            Waiting for a sample color choice.
            """
        case "demo1|w2:p1":
            """
            Fictional Reading Journal demo

            Exported three imaginary books:
              The Clockwork Orchard
              A Map of Paper Moons
              The Pocket Raincloud

            The sample export is ready to review.
            """
        case "demo2|w1:p2":
            """
            Fictional Art Notebook demo

            A sample postcard has a blue sky and two paper kites.
            Choose a dotted or striped border for the example.
            """
        case "demo2|w2:p1":
            """
            Fictional Weather Samples demo

            Example readings: sunny, breezy, light rain.
            All values are invented and describe no real location.
            """
        default:
            """
            Fictional Garden Planner demo

            Draft seed list: basil, thyme, marigold.
            The sample garden has three raised beds.
            The example layout is ready for a color preview.
            """
        }
    }

    static func gitStatus(for workspace: HerdrWorkspace) -> WorkspaceGitStatus {
        WorkspaceGitStatus(
            ok: true,
            workspaceID: workspace.id,
            branch: workspace.tokens["branch"] ?? "demo/garden-colors",
            cwd: workspace.displayPath,
            staged: [
                WorkspaceGitFile(status: "M", file: "Sources/Garden/GardenCanvas.swift"),
            ],
            unstaged: [
                WorkspaceGitFile(status: "M", file: "Sources/Garden/SeedPicker.swift"),
                WorkspaceGitFile(status: "M", file: "Sources/Reading/BookList.swift"),
            ],
            untracked: [
                "Sources/Garden/PlantCatalog.swift",
            ],
            commits: [
                WorkspaceGitCommit(hash: "111aaaa", message: "Add the fictional seed list"),
                WorkspaceGitCommit(hash: "222bbbb", message: "Arrange example garden beds"),
                WorkspaceGitCommit(hash: "333cccc", message: "Label sample plant colors"),
            ],
            error: nil
        )
    }

    static func gitDiff(file: String, section: GitFileSection) -> WorkspaceGitDiffResponse {
        WorkspaceGitDiffResponse(
            ok: true,
            file: file,
            section: section,
            diff: """
            diff --git a/\(file) b/\(file)
            index 17f4c31..92ab840 100644
            --- a/\(file)
            +++ b/\(file)
            @@ -42,6 +42,9 @@ struct GardenCanvas: View {
                 PlantTile(
                     plant: samplePlant,
            +        color: sampleColor,
            +        isSelected: $isSelected,
            +        select: selectSamplePlant,
                     label: sampleLabel
                 )
            """,
            error: nil
        )
    }

    static func skills(for workspace: HerdrWorkspace) -> SkillsResponse {
        let projectSkills = [
            ProjectSkill(
                name: "garden-layout",
                skillFilePath: "./.claude/skills/garden-layout/SKILL.md",
                scope: "project"
            ),
            ProjectSkill(
                name: "sample-review",
                skillFilePath: "./.claude/skills/sample-review/SKILL.md",
                scope: "project"
            ),
            ProjectSkill(
                name: "catalog-export",
                skillFilePath: "./.claude/skills/catalog-export/SKILL.md",
                scope: "project"
            ),
        ]
        let userSkills = [
            ProjectSkill(
                name: "demo-handoff",
                skillFilePath: "~/.codex/skills/demo-handoff/SKILL.md",
                scope: "user"
            ),
            ProjectSkill(
                name: "example-notes",
                skillFilePath: "~/.codex/skills/example-notes/SKILL.md",
                scope: "user"
            ),
        ]
        return SkillsResponse(
            ok: true,
            workspaceID: workspace.id,
            rootPath: workspace.displayPath,
            skillsDirectory: "\(workspace.displayPath)/.claude/skills",
            userSkillsDirectory: "~/.codex/skills",
            projectSkills: projectSkills,
            userSkills: userSkills,
            skills: projectSkills + userSkills,
            error: nil
        )
    }

    static func fileSearch(query: String, workspace: HerdrWorkspace) -> FileSearchResponse {
        FileSearchResponse(
            ok: true,
            workspaceID: workspace.id,
            rootPath: workspace.displayPath,
            query: query,
            files: [
                ProjectFileMatch(path: "Sources/Garden/GardenCanvas.swift"),
                ProjectFileMatch(path: "Sources/Garden/SeedPicker.swift"),
                ProjectFileMatch(path: "Tests/GardenCanvasTests.swift"),
            ].filter { query.isEmpty || $0.path.localizedCaseInsensitiveContains(query) },
            truncated: false,
            limit: 50,
            error: nil
        )
    }

    static let jiraTickets = JiraTicketsResponse(
        ok: true,
        project: "TASK",
        projects: ["TASK", "APP"],
        site: "example.atlassian.net",
        tickets: [
            JiraTicket(
                key: "TASK-101",
                projectKey: "TASK",
                title: "Choose a fictional garden layout",
                status: "In Progress",
                priority: "High",
                issueType: "Story",
                url: "https://example.atlassian.net/browse/TASK-101"
            ),
            JiraTicket(
                key: "APP-102",
                projectKey: "APP",
                title: "Label the sample weather chart",
                status: "Ready for QA",
                priority: "Medium",
                issueType: "Task",
                url: "https://example.atlassian.net/browse/APP-102"
            ),
        ],
        error: nil
    )

    private static func workspace(
        id: String,
        number: Int,
        label: String,
        path: String,
        panes: [HerdrPane],
        layouts: [HerdrLayout]
    ) -> HerdrWorkspace {
        let status = panes.min(by: { $0.agentStatus.attentionRank < $1.agentStatus.attentionRank })?.agentStatus ?? .unknown
        return HerdrWorkspace(
            workspaceID: id,
            number: number,
            label: label,
            focused: number == 1,
            paneCount: panes.count,
            tabCount: Set(panes.map(\.tabID)).count,
            activeTabID: panes.first?.tabID ?? "",
            agentStatus: status,
            tokens: ["branch": number == 1 ? "demo/garden-colors" : "main"],
            worktree: HerdrWorktree(
                repoKey: label.lowercased().replacing(" ", with: "-"),
                repoName: label,
                repoRoot: path,
                checkoutPath: path,
                isLinkedWorktree: number == 1
            ),
            tabs: Array(Set(panes.map(\.tabID))).sorted().enumerated().map { index, tabID in
                HerdrTab(
                    tabID: tabID,
                    workspaceID: id,
                    number: index + 1,
                    label: index == 0 ? "Agents" : "Tests",
                    focused: index == 0,
                    paneCount: panes.count(where: { $0.tabID == tabID }),
                    agentStatus: status
                )
            },
            panes: panes,
            layouts: layouts
        )
    }

    private static func pane(
        id: String,
        tabID: String,
        status: AgentStatus,
        title: String,
        agent: String?,
        revision: Int,
        firstSeenAt: Date? = nil,
        lastActivityAt: Date? = nil,
        workingSince: Date? = nil
    ) -> HerdrPane {
        HerdrPane(
            paneID: id,
            terminalID: "term_\(id.replacing(":", with: "_"))",
            workspaceID: String(id.split(separator: ":").first ?? "w1"),
            tabID: tabID,
            focused: id == "w1:p1",
            agentStatus: status,
            revision: revision,
            cwd: "/tmp/herdr-demo",
            foregroundCWD: nil,
            label: nil,
            title: title,
            agent: agent?.lowercased(),
            displayAgent: agent,
            terminalTitle: title,
            terminalTitleStripped: title,
            firstSeenAt: firstSeenAt,
            lastActivityAt: lastActivityAt,
            workingSince: workingSince
        )
    }

    private static func layout(workspaceID: String, tabID: String, paneIDs: [String]) -> HerdrLayout {
        let width = 120
        let paneWidth = width / max(paneIDs.count, 1)
        return HerdrLayout(
            workspaceID: workspaceID,
            tabID: tabID,
            focusedPaneID: paneIDs.first,
            zoomed: false,
            area: HerdrLayoutRect(x: 0, y: 0, width: width, height: 36),
            panes: paneIDs.enumerated().map { index, id in
                HerdrLayoutPane(
                    paneID: id,
                    focused: index == 0,
                    rect: HerdrLayoutRect(x: index * paneWidth, y: 0, width: paneWidth, height: 36)
                )
            },
            splits: []
        )
    }
}
